Themes, borders and little bits

belakm opened the pull request:
Hey there!

Due the needs of my project i couldnt extend or change React95 theme and elements in my project so i went down the rabbit hole of setting things in such a way that i could use different palettes and re-use colors in my app.

I've also created an addon for React95 storybook that changes the selected themes in the preview but as far as i can tell is not reflected in the recent storybook addition.

![_themes](https://user-images.githubusercontent.com/13392444/99993633-3f0d4a00-2db8-11eb-8b73-6ecb5a1d9a55.png)

I've also corrected the shadows such that they are exactly like in windows95.
Screenshots of windows95 that were used for reference are from this page: [https://guidebookgallery.org/screenshots/win95](https://guidebookgallery.org/screenshots/win95)

![_borders](https://user-images.githubusercontent.com/13392444/99994139-ec805d80-2db8-11eb-8064-244e4b882a66.png)

There are 2 distinct shadow patterns that i've discerned that aren't as others, these are scrollbars' and windows' borders but may also appear elsewhere. Some elements from original win95 do have different shadows from program to program though but these are usually custom elements.

# Changes:

## Theming
- Theming via styled-components themes. I've also copied around 30 themes from original React95 project.
- Border helpers in /shared-style/Border.tsx can be used to create win95 style borders. Comes in 2 styles: With border and :after with box-shadow and box-shadow only.

I've also exported types for styled-components theme used in theming so it can be re-used or extended in other styled-components projects. More on this in the **Using with styled-components themes** section.

## Components

### Button
- shadows
- added disabled Button look 

### Checkbox
- shadows
- tweaked label a bit so it centers with box better

### Cursors
- moved cursor logic to a styled-components helper function, can now be used in tandem with styled-components without the need for classes eg.

```typescript
import styled from 'styled-components'
import Cursor from '@react95/core/shared-style/Cursor'

const componentWithCursor = styled.p`
  ${Cursor('auto')} // <-- Will add proper cursor properties
`

// ...
``` 

I left class definitions in global style as they were, just replaced them with calls to Cursor helper.

### Divider

- created a Divider component for horizontal breaks. Would be nice to have a vertical one as well.

### Dropdown

- borders
- Replaced triangle image with an SVG generated element of a triangle.

### Input

- borders

### Modal

This is probably the definitive element of this UI kit - the classic Windows95 style window.
The shadows didn't match the original though so i fixed that by creating another component called Window meant only for Window background.

I've also tweaked the story examples a bit by adding content to them.
I've moved the title bar to its own file, because i find it easer to compose components in my app if i can put together Frame + Title bar without the whole Modal functionality.

I've also switched the identifier of modals in ModalProvider from 'title' to 'id' that generates with each modal.

### Panel

A Panel element, similar to Frame. Made in a way that adjoining panels do not have a double border.
I didn't get the Frame element at first as it is so simplistic and mostly just xstyled-component bindings, so i made a more streamlined version.

I've also included 'as canvas' mode that creates a distinct different background for every theme, for example such as white color was used for areas with a lot of text;

By default all Panels are wrapped in another element because that is the only way to keep a double shadow  in a scrolling container (box shadow will be rendered underneath the content scrolled. This can be disabled by setting `isScrollable={false}` prop.

Panels are by default `display: flex` and their direction attribute is used to determine how to merge shadows of inner Panels.

![image](https://user-images.githubusercontent.com/13392444/100024586-629bb900-2de7-11eb-95ad-03394b7df35b.png)

### Progress Bar

- borders

###  Tabs

- borders

### Taskbar

I've touched the story a bit and modded the elements to have colors of the themes. Also added some content to the windows in the story.

### Textarea

- borders

### ThemeProvider

Accepts either `string` or `DefaultTheme object` on theme prop. By default it uses 'win95' theme (the default win95 look).

I've moved all the themes from javascrip React95 project and used the same structure for colors as it seemed a good system. I did however omit some of the properties that would never be used at this time.

All mentioned themes are in `/ThemeProvider/themes/, where baseTheme.tsx is also defined and holds everything else that all themes share (aka zIndices and other specifics).

### Title Bar

Moved it out of the Modal so it can be reusable on a Window/Panel/Frame element without the Modal functionality - just for styling elements to look like win95 windows.

### Window

A container element that has the same shadows as win95 did.

## Using in another styled-components app

React95 theme (otherwise in `core/types/theme.d.ts`) is exported in the `/dist/_types` folder - i've tweaked the script to copy it on prepublishOnly step of the building so it gets packed with the components.

```typescript
// ThemeDefinition.d.ts
import 'styled-components'
import { R95DefaultTheme } from '@react95/core/_types/theme'

declare module 'styled-components' {

  // your stuff

  export interface DefaultTheme extends R95DefaultTheme {
    // anything else you want in your theme
  }
}
```

You can then create own theme with one of the themes as a base (for R95 theme) so typescript is happy.

```typescript
// my-theme.ts
import win95 from '@react95/core/ThemeProvider/themes/win95'
import { DefaultTheme } from 'styled-components'

const Theme: DefaultTheme = {
  ...win95,
  // your stuff here
}

export default Theme
```

and use your Theme provider normally with any other win95 theme

```typescript
import { ThemeProvider } from 'styled-components'
import { GlobalStyle, ThemeProvider as R95ThemeProvider } from '@react95/core'
import Theme from '../your/custom/Theme'
const App = () => 
  <ThemeProvider theme={Theme}>
    your default R95 theme defined in Theme file
    <GlobalStyle />
      <R95ThemeProvider theme='rose'>
        // custom content with candy theme inside
      </R95ThemeProvider>
  </ThemeProvider>
```

Here shown: win95 in the header in conjunction with rose theme for 'windows' elements:
![image](https://user-images.githubusercontent.com/13392444/100019392-40ea0400-2dde-11eb-9821-1d889f50418c.png)

## Other

There may be some refactorings i forgot to mention because they slip my mind at this time. I've worked on this for the past month on weekends and pasttime and didnt commit as much as i should so there are a lot of changes in the first commits when i was preparing the terrain. I have removed some of the @xstyled-components code as it was feeling odd to me at the time. I've refrained from it later and made elements around that, but this may cause some style breaking an xstyled app.

ggdaltoso commented:
WOW @belakm ! just WOW

Ana amazing job, man!!

I'll need some time to review everything 😂

belakm commented:
please do :)

arturbien commented:
@belakm I've noticed on attached screenshot, that there are some issues with `TitleBar` color in `millenium` and `storm` theme (TitleBar doesn't have any color).
 
It's because the theme prop called `headerBackground`  holds a value of `linear-gradient()` in these two particular themes. (most themes `headerBackground` hold simple color values like hex or rgb)

What will fix this issue is just assigning this theme prop to `background` instead of `background-color` in `TitleBar` css.
This makes TitleBar customization more flexible because this way you can set any color, any gradient or even background image as a TitleBar background :) 

Btw I really dig the theme picker you've created! 😍 I would be happy to receive a PR with this addon in our version of React95 soooo if you'd like to help out- https://github.com/arturbien/React95/issues/215 ♥️

ggdaltoso commented:
Hey, @arturbien! I'm happy to see you here and, please, help me with this review.

The difference between `@react95/core` and [react95](https://github.com/arturbien/React95) is that this project wants to be pixel perfect whereas the one from @arturbien aim to be mobile-friendly.

And I'm a little confused when you say 

> "... original React95 project."

since this is almost a year older. But, anyway...

That said, let's start with the Button.

The refactor you did place this component next to the one from `react95`. The default style, active and focused is different from what we have on Figma, don't you agree, @janneilkka?

Here is a high resolution of what I'm talking about:

![image](https://user-images.githubusercontent.com/6536985/100448108-92211e80-3090-11eb-9179-826781bd4885.png)

To help us with that, we could use something as [Chicago95 Linux Theme](https://github.com/grassmunk/Chicago95) does on it's [`CSS`](https://github.com/grassmunk/Chicago95/blob/master/Theme/Chicago95/cinnamon/cinnamon.css#L77), using [`border-image`](https://developer.mozilla.org/en-US/docs/Web/CSS/border-image). 

MDN also has an [awesome generator](https://developer.mozilla.org/pt-BR/docs/Web/CSS/Tools/Border-image_generator) to help us too.

I'll keep doing the review and break it into more comments. Stay tuned! 😂

ggdaltoso commented:
Also, @belakm, I use the [executable windows95](https://github.com/felixrieseberg/windows95) to help me when I'm developing to `@react95/core`. I hope it could help you too!

belakm commented:
@ggdaltoso I didn't know these projects arent connected, about that, im sorry 😳 i assumed this is a typescript rewrite. I found about this thru [styled-components ecosystem page](https://styled-components.com/ecosystem) where react95 is mentioned and googled it with "typescript" hoping someone did it 😅 

I've tested a few designs while working on this PR on how borders work, and i cant remember anymore to be honest why i went with border + :after with box-shadow / border and border/box-shadow variants.

Theres basically 2 ways to make double border;

- border + box-shadow
- box-shadow alone

And subvariants with using :after/:before/wrapper components etc.

Since borders arent pixel perfect (renders will interpolate the 4 outermost border pixels), we might wanna use box-shadow exclusively then.

Borders with box-shadows:

![image](https://user-images.githubusercontent.com/13392444/100519026-97b45c80-3195-11eb-8259-52b8170a412c.png)

----

I'd advise against raster images for whatever purpose unless totally neccessary like with cursors, icons or for example checkbox image - those that cant be easily recreated in SVG and are an additional load on network and bundle size. That's why i removed the triangles in scrollbars and elsewhere with @arturbien's triangle svg helpers.

**I've pushed a fix where Borders are exclusively made of box-shadows, so take a look.**

I've also simplified the Border.tsx style helper and Panel component by removing now-unnecessary stuff.

I've also fixed the detail with linear-gradient @arturbien suggested on TitleBar and it does look really good 😄 ;

![image](https://user-images.githubusercontent.com/13392444/100523817-ae69ac00-31b3-11eb-8cdd-b70983274227.png)

ggdaltoso commented:
Impressive work, @belakm 

I still think we are having some mismatching against button states and I would like to ask you to revisit then

`:focus`
| Yours        | w95             |
| ---------- |------------|
| ![focus-new](https://user-images.githubusercontent.com/6536985/100737165-90b66580-33b2-11eb-9faf-19e0e7cf850a.gif) | ![focus](https://user-images.githubusercontent.com/6536985/100736269-45e81e00-33b1-11eb-9780-d4ba373070d6.gif) |

`:active`
| Yours        | w95             |
| ---------- |------------|
| ![active-new](https://user-images.githubusercontent.com/6536985/100737848-7df06080-33b3-11eb-80d4-96aa18cf99e7.gif) |![w95-active](https://user-images.githubusercontent.com/6536985/100737876-86e13200-33b3-11eb-9f11-18afc5988c8b.gif)|

When `:active`, notice a slight difference in paddings, making the text a little down and right.

janneilkka commented:
Shouting from the sidelines here but indeed @ggdaltoso - I also noticed this mismatch in our Figma document, `:focus` does not have the black bordering as in your demonstration! I will put this into backlog.

As for the paddings excatly just like @ggdaltoso demonstrated it quite effectively, padding of the text is a funny gimmick in win95 when button pushed down. :) This is demonstrated as needed in Figma.

ggdaltoso commented:
For Checkbox, we have some mismatch with disabled text too:

![image (3)](https://user-images.githubusercontent.com/6536985/100741216-a890e800-33b8-11eb-8235-d40bbfb03db8.png)

We can easily find a disabled checkbox in "Display Properties"
![image](https://user-images.githubusercontent.com/6536985/100742498-a3806880-33b9-11eb-9d60-28f36cd2d177.png)

ggdaltoso commented:
There is also an example of a disabled Button on the comment above.

ggdaltoso commented on packages/core/components/Button/Button.tsx:
```diff
@@ -1,9 +1,54 @@
 
-const Button = Btn;
+export const StyledButton = styled.button<ButtonProps>`
+  background-color: ${({ theme }) => theme.colors.material};
+  color: ${({ theme }) => theme.colors.materialText};
+  padding: 6px 22px 6px;
```
One of the reasons that we use `@xstyled` is because we don’t need to add this callback so many times. It simplifies the code and forces us to always get values from the theme itself.

```suggestion
  background-color: material; // this will get material from theme.colors
  color: materialText; // this will get materialText from theme.colors
  padding: 6 22 6; // this will get the 6th and 22th from theme.spacing
```

ggdaltoso replied:
One of the reasons that we use `@xstyled` is because we don’t need to add this callback so many times. It simplifies the code and forces us to always get values from the theme itself.

```suggestion
  /* this will get material from theme.colors */
  background-color: material;

  /* this will get materialText from theme.colors */
  color: materialText;
  
  /* this will get the 6th and 22th from theme.spacing */
  padding: 6 22 6; 
```

belakm commented:
Closing this.

ggdaltoso commented:
Hey, @belakm! A long time to see you here!

I think I've covered some of the features you were proposing in this PR in another ones. 
I let this PR opened to guide and help me while I was breaking it into smaller pieces.

Would you like to take a time and have a look?
You're always welcome to contribute <3

belakm commented:
Hey @ggdaltoso :) Sorry but i never took the time needed to do additional work. I have made modifications and submitted a pull request after i was happy to have finished what i planned as it was for another project I was at the time working on. Should I need win95 look again I will be happy to return to this library. I wish you the best <3
